### PR TITLE
[canvaskit] Clip before applying ColorFilter so it doesn't filter beyond child bounds

### PR DESCRIPTION
**Pull request by harryterkelsen**
When a ColorFilter affects transparent black, it will expand its bounds to the entire screen, even if the `saveLayer` call is bounded. This applies a clip before applying the ColorFilter so the filter is bounded to just the child drawings.

Also fixes bug with ColorFilter being used as an ImageFilter.

Before:
canvaskit_colorfilter_bounds_before

After:
canvaskit_colorfilter_bounds_after

Fixes https://github.com/flutter/flutter/issues/88866
Fixes https://github.com/flutter/flutter/issues/144015

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/layer.dart [hunk 1]
```diff
@@ -518,10 +533,15 @@ class ColorFilterEngineLayer extends ContainerLayer
     final CkPaint paint = CkPaint();
     paint.colorFilter = filter;
 
+    paintContext.internalNodesCanvas.save();
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/layer.dart:536]: Maybe leave a comment explaining why clipping is needed (and also leave a link to the issue that this fixed).

File: lib/web_ui/lib/src/engine/canvaskit/layer.dart [hunk 2]
```diff
@@ -518,10 +533,15 @@ class ColorFilterEngineLayer extends ContainerLayer
...
+    paintContext.internalNodesCanvas
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/layer.dart:537]: Can we leave a TODO to remove the clip for filters that do not affect transparent black? Clips can be expensive.

**Review by yjbanov — APPROVED** [review]

File: lib/web_ui/lib/src/engine/canvaskit/layer.dart [hunk 3]
```diff
@@ -409,12 +411,23 @@ class ImageFilterEngineLayer extends ContainerLayer
     childMatrix.translate(_offset.dx, _offset.dy);
     prerollContext.mutatorsStack
         .pushTransform(Matrix4.translationValues(_offset.dx, _offset.dy, 0.0));
+    final CkManagedSkImageFilterConvertible convertible;
```
[the remark this answers is not in the document]
Comment (harryterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/layer.dart:414]: It's only used in `preroll`. In `paint` we use the `filter` directly

**Comment by zanderso** [comment]
This change had SkiaGold diffs in two tests on the roll to the framework. https://flutter-gold.skia.org/search?crs=github&issue=148147&patchsets=8&positive=true&untriaged=false They looked correct, and I accepted them.

**Pull request merged**